javascripts/2048.js

```diff
@@ -1,32 +1,276 @@
 $(document).ready(function() {
   console.log('ready!');
+  arrayInitialize();
+  initializeGame();
   $('body').keydown(function(event){
     var arrow_keys = [37, 38, 39, 40];
+
     if(arrow_keys.indexOf(event.which) > -1) {
-      var tile = $('.tile');
-      moveTile(tile, event.which);
+      // var tile = $('.tile');
+      moveTile(event.which);
       event.preventDefault();
     }
   })
 })
 
-function moveTile(tile, direction) {
+function pop(tile) {
+  $(tile).addClass("popper")
+  .on("animationend",function() {
+    $(this).removeClass("popper");
+  });
+}
+
+
+ //keeps track of where on the board has a tile
+var tile_array = [];
+//set up for 2D array
+function arrayInitialize(){
+  for(var i = 0; i < 4; i++) {
+    var first_array = [];
+    tile_array.push(first_array);
+    for(var j = 0; j < 4; j++) {
+      setUndefinedTile(i, j);
+      }
+    }
+}
+
+function setUndefinedTile(col, row) {
+  tile_array[col][row] = undefined;
+}
+
+function setTileInArray(col, row, tile) {
+  tile_array[col][row] = tile;
+}
+
+
+// Creates 2 tiles at the start of the game
+function initializeGame() {
+  createTile();
+  createTile();
+}
+
+// Random number generated between 0 and 3
+function rando_num(){
+  return Math.floor(Math.random() * 4);
+}
+
+// See if a tile already exists at a particular column and row
+function checkLocation(column, row) {
+  if (tile_array[column][row] !== undefined) {
+    column = rando_num();
+    row = rando_num();
+    checkLocation(column, row);
```

Comment:
However, recursion can be tricky. There's a subtle bug that will make this not work, as you never return the right location if you go into your if block. One of the things you might look for is that your always return when calling a function recursively within itself.

Here's how you could fix it.
```
function checkLocation(column, row) {
  if (tile_array[column][row] === undefined) {
    return [column, row]
  } else {
    column = rando_num();
    row = rando_num();
    return checkLocation(column, row);
  }
}
```